pkg/file/redundancy/getter/getter.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package getter
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/klauspost/reedsolomon"
+)
+
+// decoder is a private implementation of storage.Getter
+// if retrieves children of an intermediate chunk potentially using erasure decoding
+// it caches sibling chunks if erasure decoding started already
+type decoder struct {
+	fetcher    storage.Getter  // network retrieval interface to fetch chunks
+	putter     storage.Putter  // interface to local storage to save reconstructed chunks
+	addrs      []swarm.Address // all addresses of the intermediate chunk
+	inflight   []atomic.Bool   // locks to protect wait channels and RS buffer
+	cache      map[string]int  // map from chunk address shard position index
+	waits      []chan struct{} // wait channels for each chunk
+	rsbuf      [][]byte        // RS buffer of data + parity shards for erasure decoding
+	ready      chan struct{}   // signal channel for successful retrieval of shardCnt chunks
+	shardCnt   int             // number of data shards
+	parityCnt  int             // number of parity shards
+	wg         sync.WaitGroup  // wait group to wait for all goroutines to finish
+	mu         sync.Mutex      // mutex to protect buffer
+	fetchedCnt atomic.Int32    // count successful retrievals
+	cancel     func()          // cancel function for RS decoding
+	remove     func()          // callback to remove decoder from decoders cache
+}
+
+type Getter interface {
+	storage.Getter
+	io.Closer
+}
+
+// New returns a decoder object used tos retrieve children of an intermediate chunk
+func New(addrs []swarm.Address, shardCnt int, g storage.Getter, p storage.Putter, strategy Strategy, strict bool, fetchTimeout time.Duration, remove func()) Getter {
+	ctx, cancel := context.WithCancel(context.Background())
+	size := len(addrs)
+	if fetchTimeout == 0 {
+		fetchTimeout = 30 * time.Second
```

Comment:
defaults should not defined in the constructor, import it from the retrieval pkg